### PR TITLE
[base] Renamed XXXThreadPool for better understanding. Removed tautology namespaces.

### PR DESCRIPTION
**Pull request by vng**

### CODE REVIEW

File: base/base_tests/thread_pool_tests.cpp [hunk 1]
```diff
@@ -89,23 +90,18 @@ namespace
...
 
-  for (size_t i = 0; i < tasks.size(); ++i)
-    pool.PushBack(tasks[i]);
+  for (int i = 0; i < TASK_COUNT - 1; ++i)
+    pool.PushBack(new CancelTestTask(true));
 
   // CancelTastTask::Do method should not be called for last task
-  tasks.back()->Cancel();
-  tasks.clear();
+  auto * p = new CancelTestTask(false);
```
Comment (biodranik) [inline base/base_tests/thread_pool_tests.cpp:102]: Is it properly deleted? unique_ptr?

File: base/workers_pool.hpp [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+        }
+      });
+    }
+  }
+
+  void Add(std::unique_ptr<Task> && task)
+  {
+    std::unique_lock lock(m_mtx);
+    m_tasks.push(std::move(task));
+    m_cond.notify_all();
```
Comment (biodranik) [inline base/workers_pool.hpp:65]: Why all threads should be awakened for one task?
```suggestion
    m_cond.notify_one();
```

File: base/workers_pool.hpp [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+{
+public:
+  class Task
+  {
+  public:
+    virtual ~Task() = default;
+    virtual void Do() = 0;
+  };
+
+  explicit WorkersPool(size_t threadsNum)
```
Comment (biodranik) [inline base/workers_pool.hpp:25]: Why another pool? Because the old one was single-threaded?

File: base/workers_pool.hpp [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    std::unique_lock lock(m_mtx);
+    m_tasks.push(std::move(task));
+    m_cond.notify_all();
+  }
+
+  void Finish()
+  {
+    {
+      std::unique_lock lock(m_mtx);
+      m_cond.wait(lock, [this]() { return m_tasks.empty(); });
```
Comment (biodranik) [inline base/workers_pool.hpp:72]: Finish() is blocking here, why?

**Review by biodranik — APPROVED** [review]

File: base/base_tests/thread_pool_tests.cpp (hunk 1 above)
Comment (vng) [reply base/base_tests/thread_pool_tests.cpp:102]: base::ThreadPool is designed that tasks are deleted by separate finish function.

File: base/workers_pool.hpp (hunk 2 above)
Comment (vng) [reply base/workers_pool.hpp:65]: We can wait in worker and in Finish

File: base/workers_pool.hpp (hunk 3 above)
Comment (vng) [reply base/workers_pool.hpp:25]: ThreadPool is a bit fancy and designed for work-cancel-stop.

File: base/workers_pool.hpp (hunk 4 above)
Comment (vng) [reply base/workers_pool.hpp:72]: Finish should complete all tasks.

Comment (biodranik) [reply base/workers_pool.hpp:72]: That's an interesting approach. What if tasks are too long, and the finish is called from another critical thread? Main one?

Comment (vng) [reply base/workers_pool.hpp:73]: That is a point. It is like simple tbb::parallel_for.

**Comment by vng** [comment]
Oh, I didn't find base::thread_pool::computational::ThreadPool :)

**Review by biodranik — APPROVED** [review]
OMG how much custom threading code we have. It would be great to use more efficient system pools like GCD if possible.

**Pull request merged**